### ISSUE TITLE
Setting CAN IDs and Firmware through Linux

### ISSUE THREAD

**Issue by thatrobotguy** [CrossTheRoadElec/Phoenix-Linux-SocketCAN-Example#13]
I see that there is a Windows Application for configuring CAN bus devices connected to a RoboRio, but I do not see how to format CTRE CAN devices through Linux. I have my control system configured with a CanAble CAN bus adapter driving multiple Talon SRXs, a CANifier and a Pigeon. Is there a C++ API call to load firmware to a device? Is there a Linux or Web application to connect to the network? What Linux options are there for formatting/configuring CTRE devices?

I am able to unplug all of the devices on the CAN network except for 1 so as to make sure the correct device gets updated.
I am using Ubuntu 1804.

**Comment by jeholliday** [comment]
I believe the best way to contact CTRE is through their support email address: support@ctr-electronics.com. They have been very responsive to me by using that method.

You are correct that there is no non-windows utility for flashing the firmware of their CAN devices. However, by following the [directions here](https://github.com/CrossTheRoadElec/Phoenix-Linux-SocketCAN-Example#connecting-to-diagnostics-server), you can use the Phoenix Tuner to flash CAN devices connected to a linux machine. It's kind of finicky; I've had to create an admin account with no password (the same a s the robo-rio) to make it recognize my Jetson, but it works. An easier solution is to use their HERO board and its utility to flash CAN devices, but it still requires Windows.

**Comment by JCaporuscio** [comment]
@jeholliday is correct.

Also, there is an open-source version of our flashing utility that you can see here:
https://github.com/CrossTheRoadElec/Phoenix-diagnostics-client
This still requires the use of the Phoenix diagnostics server on your CAN platform.

This can be a good starting point if you want to write your own utility to interact with diag-server's http interface.
You could also try compiling and running the open-source utility using mono on a linux system.

**Comment by thatrobotguy** [comment]
Responding to what @jeholliday said:

Looking at the link you provided, I am reading it as saying that there is still a Windows computer required to connect to the control system. It says that "Make sure your device is connected to the same Wi-Fi network as the windows PC you want to run Tuner on."
It sounds like I have to have a Windows Machine install the software onto my Linux Laptop rather than download it to my Linux machine then install it. It is also my understanding that the software that gets installed through this method is supposed to be the most up-to-date version.
Is this then a utility I can start from CLI or is it an always-running background service?
It looks like the steps provided create an `ssh` session to the linux machine and then installs the software.
I have to say, I was not entirely sure what that section was about, but I understand better why it is there. I would have explained a little more than just "Connecting to Diagnostics Server" because, if I understand correctly, it is actually installing the CTRE software for interfacing with the CAN network.
One more question: this still works through the CanAble adapter, correct?

Responding to what @JCaporuscio said

The link you provided says that the Phoenix Tuner is the most up-to-date software updater, not the software directly discussed/provided in the link. Is this simply because it is most up-to-date **for the roborio**? I want to make sure that the software I use for updating my devices will work for an arbitrarily long amount of time. Also, I am not well-versed in the mono project, but I have heard of it before. You say "This still requires the use of the Phoenix diagnostics server on your CAN platform." It sounds like there is a Client and a Server required to update the CAN devices. I only have the CAN network, the CanAble, and my linux laptop. Would the server and client both run on my Linux laptop?

Also, it sounds like you are suggesting 2 different methods for flashing devices on the network. Could you elaborate on those 2 methods? I am not sure I see the separation: the link you provided is a C# project, but I don't see where the second solution is coming from.

Basically, it is not clear to me which path to take, considering **mono** sounds highly-dependent on the platform and the other solution may not work on a non-Jetson device.
In other words, I don't see a clear solution provided.

**Comment by vatbrain** [comment]
I would very much like to see this supported natively on linux (no windows pc required) as well.
A simple text only app would suffice.

**Comment by JCaporuscio** [comment]
We currently don't have plans for officially supporting Linux.  This may be something we add in the future (possible a command-line utility), but there's no kind of ETA or plans to share.

That being said, there is now an *unofficial* linux GUI that's available.
https://github.com/TytanRock/PenguinTuner/releases

We don't provide any support for this utility but it provides some of the base features of Phoenix Tuner (like flashing firmware or changing device IDs).

@thatrobotguy I realize your comment is a bit old, but I wanted to address it for posterity.
You're correct that there are two pieces to Phoenix Tuner: Tuner itself which is the GUI, and the Diagnostics Server which handles CAN communication/data and provides the HTTP API that Tuner uses.

The Diagnostic Server needs to run on whichever platform has the physical CAN communication.  So in the case of a roboRIO or raspberry pi it needs to run onboard these devices.  In the case of a Linux (or Windows) machine with the CAN bus connected, the Diag Server needs to run on the Linux box itself.

We don't currently have a button in Tuner to deploy to a generic Linux machine, but you can run the server on your Linux computer manually.  You would write a basic C++ app with the Linux x64 Phoenix libraries (similar to the SocketCAN example) and call `c_SetPhoenixDiagnosticsStartTime(0);` from ctre/phoenix/cci/Unmanaged_CCI.h
Running this app would run the diagnostics server, and then you would enter your machine's IP address (or localhost if using PenguinTuner) into whichever GUI you're using.

**Comment by vatbrain** [comment]
Something like this seems to work to cause a Diagnostic Server to be run on a Raspberry Pi:
```
////////////////////////////////////////
//
// Minimal app to cause a Phoenix diagnostic server to be run.
//
////////////////////////////////////////

#define Phoenix_No_WPI // remove WPI dependencies
#include "ctre/phoenix/platform/Platform.h"
#include "ctre/phoenix/cci/Unmanaged_CCI.h"
#include <string>
#include <chrono>
#include <thread>
#include <unistd.h>

using namespace ctre::phoenix;
using namespace ctre::phoenix::platform;

/** simple wrapper for code cleanup */
void sleepApp(int ms)
{
        std::this_thread::sleep_for(std::chrono::milliseconds(ms));
}

int main() {
        std::string interface;
        interface = "can0";
        ctre::phoenix::platform::can::SetCANInterface(interface.c_str());

        c_SetPhoenixDiagnosticsStartTime(0); // we will use the diag server this application installs

        while (true) {
                sleepApp(20);
        }
        return 0;
}

**Comment by BraedenCu** [comment]
Using the above example, the Diagnostics Server appears to ignore the interface name passed to the SetCANInterface(...) and/or c_SetCANInterface(..). Is there a way to pass the CAN interface name to the Diagnostic Server other than 'can0'?

**Comment by JCaporuscio** [comment]
With the current release Diagnostic Server is hard-coded to can0.  There's a change coming that will make the Diag Server that gets compiled into a user program correctly use the interface passed in from SetCANInterface.  

This likely won't make it into this week's release as we're focusing on FRC for their kickoff, but it will likely make it into the next release after.
